File: job-finder-BE/server/src/modules/generator/ai/inference-client.ts
/**
 * Unified AI inference client via LiteLLM proxy.
 *
 * Replaces AgentManager by routing all AI requests through the LiteLLM proxy,
 * which handles provider selection, fallbacks, retries, and budget tracking.
 *
 * Exposes the same interface as the old AgentManager so callers can swap
 * with minimal changes.
 *
 * Model auto-resolution: When a `claude-document` request silently falls back
 * to Gemini (indicating a stale/retired model ID or inaccessible model), the
 * client detects the degradation and probes the Anthropic API directly for the
 * latest accessible Claude model. On success, registers the working model in
 * LiteLLM via its admin API so subsequent requests route to it automatically.
 */

import type { Logger } from 'pino'
import { logger } from '../../../logger'

// ── Task Router ─────────────────────────────────────────────────────────────

const TASK_MODEL_MAP: Record<string, string> = {
  extraction: 'local-extract',
  analysis: 'local-extract',
  document: 'claude-document',
  chat: 'claude-document',
}

const DEFAULT_MODEL = 'gemini-general'

function getModelForTask(taskType: string, useLocal = true): string {
  const model = TASK_MODEL_MAP[taskType] ?? DEFAULT_MODEL
  if (!useLocal && model.startsWith('local-')) return DEFAULT_MODEL
  return model
}

// ── Model Resolver ──────────────────────────────────────────────────────────

/**
 * Process-wide probe state. Tracks whether we've already probed Anthropic
 * and registered a working model with LiteLLM.
 */
let probeInProgress: Promise<string | null> | null = null
let lastProbeAtMs = 0
const PROBE_COOLDOWN_MS = 5 * 60 * 1000 // 5 minutes

/** @internal Reset resolved model state (for tests only) */
export function _resetResolvedModel(): void {
  probeInProgress = null
  lastProbeAtMs = 0
}

/**
 * Known Claude model IDs to probe, newest first.
 * Sonnet preferred (best quality for document generation), then Haiku (fallback).
 * Generalized names (no date suffix) resolve to the latest patch in that family.
 */
const CLAUDE_CANDIDATES = [
  'claude-sonnet-4-7',
  'claude-sonnet-4-6',
  'claude-sonnet-4-5',
  'claude-sonnet-4-0',
  'claude-haiku-4-5',
  'claude-haiku-4-5-20251001',
]

/**
 * Probe Anthropic directly for the latest working Claude model accessible
 * with the current API key. Tries each candidate with a minimal 1-token request.
 *
 * On success, registers the working model as a new `claude-document` deployment
 * in LiteLLM via its admin API, so subsequent requests route to it automatically.
 *
 * Returns the working model ID (e.g., "claude-sonnet-4-6") or null.
 * Deduplicates concurrent calls — only one probe runs at a time.
 */
async function probeWorkingClaudeModel(
  litellmBaseUrl: string,
  litellmApiKey: string,
  anthropicApiKey: string,
  log: Logger
): Promise<string | null> {
  // Deduplicate: if a probe is already running, wait for it
  if (probeInProgress) return probeInProgress

  probeInProgress = (async () => {
    // ── Step 1: Find a working model via direct Anthropic API call ──────────
    let workingModel: string | null = null

    for (const candidate of CLAUDE_CANDIDATES) {
      try {
        const response = await fetch('https://api.anthropic.com/v1/messages', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': anthropicApiKey,
            'anthropic-version': '2023-06-01',
          },
          body: JSON.stringify({
            model: candidate,
            max_tokens: 1,
            messages: [{ role: 'user', content: '.' }],
          }),
          signal: AbortSignal.timeout(15_000),
        })

        if (response.ok) {
          workingModel = candidate
          log.info({ model: candidate }, 'Claude model probe succeeded via Anthropic API')
          break
        }

        const body = await response.text().catch(() => '')
        log.debug({ model: candidate, status: response.status, body: body.slice(0, 100) }, 'Claude model probe failed')
      } catch (err) {
        log.debug({ model: candidate, err }, 'Claude model probe error')
      }
    }

    if (!workingModel) {
      log.error('All Claude model probes failed — Claude will be unavailable until API key or model access is updated')
      return null
    }

    // ── Step 2: Register working model in LiteLLM via admin API ────────────
    try {
      const headers: Record<string, string> = { 'Content-Type': 'application/json' }
      if (litellmApiKey) {
        headers.Authorization = `Bearer ${litellmApiKey}`
      }

      const response = await fetch(`${litellmBaseUrl}/model/new`, {
        method: 'POST',
        headers,
        body: JSON.stringify({
          model_name: 'claude-document',
          litellm_params: {
            model: `anthropic/${workingModel}`,
            api_key: anthropicApiKey,
          },
        }),
        signal: AbortSignal.timeout(10_000),
      })

      if (response.ok) {
        log.info(
          { resolvedModel: workingModel },
          'Registered working Claude model in LiteLLM — future claude-document requests will use it'
        )
      } else {
        const body = await response.text().catch(() => '')
        log.warn(
          { status: response.status, body: body.slice(0, 200) },
          `Found working model ${workingModel} but failed to register in LiteLLM — restart LiteLLM to apply`
        )
      }
    } catch (err) {
      log.warn({ err, model: workingModel }, 'Failed to register model in LiteLLM (non-fatal)')
    }

    return workingModel
  })()

  try {
    return await probeInProgress
  } finally {
    probeInProgress = null
  }
}

// ── Types ───────────────────────────────────────────────────────────────────

type AgentExecutionResult = {
  output: string
  agentId: string
  model: string | undefined
}

export class InferenceError extends Error {}

class NoAgentsAvailableError extends Error {
  constructor(
    message: string,
    readonly taskType: string,
    readonly triedAgents: string[]
  ) {
    super(message)
  }
}

// ── Inference Client ────────────────────────────────────────────────────────

export class InferenceClient {
  private readonly baseUrl: string
  private readonly apiKey: string
  private readonly anthropicApiKey: string
  private readonly timeoutMs: number

  /** When false, local-* models are rerouted to the default cloud model. */
  useLocalModels = true

  constructor(private readonly log: Logger = logger) {
    this.baseUrl = (process.env.LITELLM_BASE_URL || 'http://litellm:4000').replace(/\/v1\/?$/, '')
    this.apiKey = process.env.LITELLM_MASTER_KEY || ''
    this.anthropicApiKey = process.env.ANTHROPIC_API_KEY || ''
    // LITELLM_TIMEOUT is in seconds (matching Python client); convert to ms
    this.timeoutMs = Number(process.env.LITELLM_TIMEOUT || '120') * 1000
  }

  /**
   * Execute an AI task via LiteLLM proxy.
   * Drop-in replacement for AgentManager.execute().
   */
  async execute(
    taskType: string,
    prompt: string,
    modelOverride?: string,
    options: { max_tokens?: number; temperature?: number; systemPrompt?: string } = {}
  ): Promise<AgentExecutionResult> {
    const model = modelOverride || getModelForTask(taskType, this.useLocalModels)

    const result = await this.callLitellm(model, taskType, prompt, options)

    // Detect silent fallback: requested Claude alias, got Gemini back.
    // Probe Anthropic directly for a working model and register it in LiteLLM.
    // Respects cooldown to avoid hammering on sustained Claude outages.
    const requestedAlias = (modelOverride || getModelForTask(taskType, this.useLocalModels)) === 'claude-document'
    if (requestedAlias && this.isGeminiModel(result.model)) {
      this.log.warn(
        { requestedModel: model, actualModel: result.model, taskType },
        'Claude request silently fell back to Gemini'
      )

      if (!this.anthropicApiKey) {
        this.log.debug('ANTHROPIC_API_KEY not available — skipping Claude model probe')
      } else {
        const now = Date.now()
        if (now - lastProbeAtMs >= PROBE_COOLDOWN_MS) {
          lastProbeAtMs = now
          probeWorkingClaudeModel(this.baseUrl, this.apiKey, this.anthropicApiKey, this.log)
            .catch(() => { /* non-fatal */ })
        }
      }
    }

    return result
  }

  private async callLitellm(
    model: string,
    taskType: string,
    prompt: string,
    options: { max_tokens?: number; temperature?: number; systemPrompt?: string }
  ): Promise<AgentExecutionResult> {
    const url = `${this.baseUrl}/v1/chat/completions`

    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs)

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model,
          messages: [
            ...(options.systemPrompt ? [{ role: 'system', content: options.systemPrompt }] : []),
            { role: 'user', content: prompt },
          ],
          max_tokens: options.max_tokens ?? 8192,
          temperature: options.temperature ?? 0.7,
        }),
        signal: controller.signal,
      })

      if (!response.ok) {
        const body = await response.text().catch(() => '')

        if (response.status === 429) {
          throw new InferenceError(
            'AI provider rate limit exceeded. Please try again later.'
          )
        }

        if (response.status === 502 || response.status === 503) {
          throw new NoAgentsAvailableError(
            `All LiteLLM providers unavailable for model ${model}: ${body}`,
            taskType,
            [model]
          )
        }

        throw new InferenceError(
          `AI generation failed (HTTP ${response.status}): ${body.slice(0, 200)}`
        )
      }

      const data = await response.json() as {
        choices: Array<{ message: { content: string } }>
        model?: string
        usage?: { total_tokens?: number }
      }

      const output = data.choices?.[0]?.message?.content || ''
      const actualModel = data.model || model

      this.log.info(
        { taskType, model: actualModel, tokens: data.usage?.total_tokens },
        'LiteLLM call succeeded'
      )

      return { output, agentId: `litellm:${model}`, model: actualModel }
    } catch (err) {
      if (err instanceof InferenceError || err instanceof NoAgentsAvailableError) {
        throw err
      }

      if (err instanceof DOMException && err.name === 'AbortError') {
        throw new InferenceError(
          `AI generation timed out after ${this.timeoutMs / 1000} seconds`
        )
      }

      const msg = err instanceof Error ? err.message : 'Unknown error'
      this.log.error({ err, model }, 'LiteLLM request failed')
      throw new InferenceError(`AI generation failed: ${msg}`)
    } finally {
      clearTimeout(timeout)
    }
  }

  private isClaudeModel(model?: string): boolean {
    if (!model) return false
    const m = model.toLowerCase()
    return m === 'claude-document' || m.includes('claude') || m.includes('anthropic')
  }

  private isGeminiModel(model?: string): boolean {
    if (!model) return false
    return model.toLowerCase().includes('gemini')
  }

  /**
   * Stream a chat completion via LiteLLM proxy (SSE).
   * Used by ChatService for real-time streaming.
   */
  async *streamChat(
    messages: Array<{ role: string; content: string }>,
    systemPrompt?: string,
    model?: string,
    options: { max_tokens?: number } = {}
  ): AsyncGenerator<string> {
    const resolvedModel = model || getModelForTask('chat', this.useLocalModels)

    const url = `${this.baseUrl}/v1/chat/completions`

    const body: Record<string, unknown> = {
      model: resolvedModel,
      messages: [
        ...(systemPrompt ? [{ role: 'system', content: systemPrompt }] : []),
        ...messages,
      ],
      max_tokens: options.max_tokens ?? 1024,
      stream: true,
    }

    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs)

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      })

      if (!response.ok || !response.body) {
        const body = await response.text().catch(() => '')
        throw new InferenceError(`LiteLLM stream failed: HTTP ${response.status}: ${body.slice(0, 200)}`)
      }

      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue
          const data = line.slice(6).trim()
          if (data === '[DONE]') return

          try {
            const parsed = JSON.parse(data) as {
              choices: Array<{ delta: { content?: string } }>
            }
            const content = parsed.choices?.[0]?.delta?.content
            if (content) yield content
          } catch (parseErr) {
            this.log.debug({ chunkLength: data.length, err: parseErr }, 'Skipping malformed SSE chunk')
          }
        }
      }
    } catch (err) {
      if (err instanceof InferenceError) throw err
      if (err instanceof DOMException && err.name === 'AbortError') {
        throw new InferenceError(`Chat stream timed out after ${this.timeoutMs / 1000} seconds`)
      }
      throw err
    } finally {
      clearTimeout(timeout)
    }
  }
}

File: job-finder-worker/src/job_finder/ai/inference_client.py
"""Unified AI inference client via LiteLLM proxy.

Replaces AgentManager by routing all AI requests through the LiteLLM proxy,
which handles provider selection, fallbacks, retries, and budget tracking.

The InferenceClient exposes the same `.execute()` interface as the old
AgentManager so callers can swap with minimal changes.
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from typing import Optional

from openai import OpenAI, APIConnectionError, APIStatusError, APITimeoutError

from job_finder.ai.task_router import get_model_for_task
from job_finder.exceptions import (
    AIProviderError,
    NoAgentsAvailableError,
    QuotaExhaustedError,
    TransientError,
)

logger = logging.getLogger(__name__)

# LiteLLM proxy defaults (overridable via env)
_DEFAULT_BASE_URL = "http://litellm:4000"
_DEFAULT_TIMEOUT = 120


@dataclass
class AgentResult:
    """Result from an inference execution.

    Kept identical to the old agent_manager.AgentResult so callers don't change.
    """

    text: str
    agent_id: str
    model: str


class InferenceClient:
    """Thin wrapper around OpenAI SDK pointed at LiteLLM proxy.

    Provides the same `.execute()` interface as the old AgentManager:
        result = client.execute(task_type="extraction", prompt="...", max_tokens=2048)
        result.text  # response content

    LiteLLM handles:
    - Provider routing (Claude, Gemini, Ollama)
    - Fallback chains (configured in litellm-config.yaml)
    - Retries and timeouts
    - Budget tracking
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        api_key: Optional[str] = None,
        timeout: Optional[int] = None,
    ):
        """Initialize the inference client.

        Args:
            base_url: LiteLLM proxy origin (default: http://litellm:4000)
            api_key: LiteLLM master key (default: from LITELLM_MASTER_KEY env)
            timeout: Request timeout in seconds (default: 120)
        """
        self._base_url = base_url or os.getenv("LITELLM_BASE_URL", _DEFAULT_BASE_URL)
        self._api_key = api_key or os.getenv("LITELLM_MASTER_KEY", "")
        self._timeout = timeout or int(os.getenv("LITELLM_TIMEOUT", str(_DEFAULT_TIMEOUT)))

        self._client = OpenAI(
            base_url=f"{self._base_url.rstrip('/')}/v1",
            api_key=self._api_key,
            timeout=self._timeout,
        )

    def execute(
        self,
        task_type: str,
        prompt: str,
        model_override: Optional[str] = None,
        max_tokens: int = 1000,
        temperature: float = 0.7,
        scope: Optional[str] = None,  # Accepted for API compat, unused
    ) -> AgentResult:
        """Execute an AI task via LiteLLM proxy.

        Drop-in replacement for AgentManager.execute(). The `scope` parameter
        is accepted for API compatibility but ignored — LiteLLM handles
        scoping via its own virtual keys.

        Args:
            task_type: Task type ("extraction", "analysis", "document", "chat")
            prompt: The prompt to send
            model_override: Override the task router's model selection
            max_tokens: Maximum tokens in response
            temperature: Sampling temperature
            scope: Ignored (kept for API compatibility)

        Returns:
            AgentResult with response text and metadata

        Raises:
            NoAgentsAvailableError: LiteLLM returned 503 (all providers down)
            QuotaExhaustedError: LiteLLM returned 429 (rate/budget limit)
            TransientError: Timeout or temporary connection failure
            AIProviderError: Other API errors
        """
        model = model_override or get_model_for_task(task_type)

        try:
            response = self._client.chat.completions.create(
                model=model,
                messages=[{"role": "user", "content": prompt}],
                max_tokens=max_tokens,
                temperature=temperature,
            )

            text = response.choices[0].message.content or ""
            actual_model = response.model or model

            logger.info(
                "LiteLLM call succeeded: task=%s model=%s tokens=%s",
                task_type,
                actual_model,
                getattr(response.usage, "total_tokens", "?"),
            )

            return AgentResult(
                text=text,
                agent_id=f"litellm:{model}",
                model=actual_model,
            )

        except APITimeoutError as e:
            raise TransientError(
                f"LiteLLM request timed out after {self._timeout}s",
                provider="litellm",
            ) from e

        except APIConnectionError as e:
            raise TransientError(
                f"Could not connect to LiteLLM proxy at {self._base_url}: {e}",
                provider="litellm",
            ) from e

        except APIStatusError as e:
            status = e.status_code
            body = str(e.body) if e.body else str(e)

            if status == 429:
                raise QuotaExhaustedError(
                    f"LiteLLM rate/budget limit: {body}",
                    provider="litellm",
                    reset_info="check LiteLLM budget settings",
                ) from e

            if status in (502, 503):
                raise NoAgentsAvailableError(
                    f"All LiteLLM providers unavailable for model {model}: {body}",
                    task_type=task_type,
                    tried_agents=[model],
                ) from e

            raise AIProviderError(f"LiteLLM API error (HTTP {status}): {body}") from e

        except Exception as e:
            raise AIProviderError(f"Unexpected error calling LiteLLM: {e}") from e
